### PR TITLE
adapterOverrides in Reconcilable interface

### PR DESCRIPTION
**Pull request by tzununbekov**
An initial set of adapter parameter overrides that includes Kn Service visibility scope switch and Resource Requests and Limits. 
Draft until testing is complete.
Discussed in triggermesh/triggermesh#705.

### CODE REVIEW

**Comment by tzununbekov** [comment]
Note: maybe we should rename `adapterOverrides` to just `adapter` in CRDs since it does not bring any additional value or clarity for users, i.e.
```
spec:
  adapterOverrides:
    public: true
```
to
```
spec:
  adapter:
    public: true
```

File: pkg/apis/common/v1alpha1/object_interfaces.go [hunk 1]
```diff
@@ -39,6 +39,9 @@ type Reconcilable interface {
 
 	// GetStatusManager returns a manager for the component's status.
 	GetStatusManager() *StatusManager
+
+	// GetAdapterOverrides returns the adapter overrides.
+	GetAdapterOverrides() *AdapterOverrides
```
Comment (antoineco) [inline pkg/apis/common/v1alpha1/object_interfaces.go:44]: I was about to suggest that we turn this into its own interface, so that we take a Go-_ish_ behaviour-driven approach and allow components to selectively implement overrides or not.
But I see you already implemented the method everywhere (crazy you!) so my suggestion suddenly becomes less relevant.

Still, if we consider that a type can be reconcilable _without_ having to override adapter settings, it might make sense to think about this as an additional `AdapterConfigurable` interface-driven behavior of some sort. Even if we want to encourage all integrations to support it at the end of the day, there might be exceptions.
The very important difference with `Reconcilable` is that, if a component doesn't implement an optional interface, things can **work** nonetheless (which is the case here).

An example of this is `EventSource` below, which can _optionally_ be implemented to expose some information about CE source and types.  In practice, all `source` types should implement it, right?
Nope! We already found a case of Source that will _not_ implement it with @odacremolbap's new `CloudEventsSource`.

File: pkg/apis/sources/v1alpha1/awscloudwatch_types.go [hunk 2]
```diff
@@ -46,6 +46,10 @@ var (
 
 // AWSCloudWatchSourceSpec defines the desired state of the event source.
 type AWSCloudWatchSourceSpec struct {
+	// Adapter spec overrides parameters.
+	// +optional
+	AdapterOverrides *v1alpha1.AdapterOverrides `json:"adapterOverrides,omitempty"`
```
Comment (antoineco) [inline pkg/apis/sources/v1alpha1/awscloudwatch_types.go:51]: nit: (applies to all Source types)
All `flow`, `extensions` and `router` types have this at the _end_ of their spec. I think it makes a lot more sense that way in terms of readability (at the end) because the type-specific attributes remain at the beginning.
Also, it's nice to be consistent in the way we do things, as a general good practice.

File: pkg/apis/targets/v1alpha1/alibabaoss_types.go [hunk 3]
```diff
@@ -44,6 +44,10 @@ var (
 
 // AlibabaOSSTargetSpec holds the desired state of the AlibabaOSSTarget.
 type AlibabaOSSTargetSpec struct {
+	// Adapter spec overrides parameters.
+	// +optional
+	AdapterOverrides *v1alpha1.AdapterOverrides `json:"adapterOverrides,omitempty"`
```
Comment (antoineco) [inline pkg/apis/targets/v1alpha1/alibabaoss_types.go:49]: nit: (applies to all Target types)
All `flow`, `extensions` and `router` types have this at the _end_ of their spec. I think it makes a lot more sense that way in terms of readability (at the end) because the type-specific attributes remain at the beginning.
Also, it's nice to be consistent in the way we do things, as a general good practice.

File: pkg/reconciler/adapter.go [hunk 4]
```diff
@@ -146,7 +146,11 @@ func commonAdapterDeploymentOptions(rcl v1alpha1.Reconcilable) []resource.Object
 		}
 	}
 
-	return objectOptions
+	overrides := rcl.GetAdapterOverrides()
+	if overrides == nil {
+		return objectOptions
+	}
+	return append(objectOptions, adapterOverrideOptions(overrides))
```
Comment (antoineco) [inline pkg/reconciler/adapter.go:153]: To echo one of my previous comments, if we were to make this behavior optional (and I think it logically makes sense to consider it optional), that logic could look like:

```golang
	if cfbl, canConfigureAdapter := rcl.(v1alpha1.AdapterConfigurable); canConfigureAdapter {
		if overrides := cfbl.GetAdapterOverrides(); overrides != nil {
				return append(objectOptions, adapterOverrideOptions(overrides))
		}
	}

	return objectOptions
```

File: pkg/reconciler/resource/podspecable.go [hunk 5]
```diff
@@ -159,3 +159,10 @@ func ConfigMapMount(name, target, cmName, cmKey string) ObjectOption {
 		})
 	}
 }
+
+func ContainerResources(resources *corev1.ResourceRequirements) ObjectOption {
+	return func(object interface{}) {
+		container := firstContainer(object)
+		container.Resources = *resources
+	}
+}
```
Comment (antoineco) [inline pkg/reconciler/resource/podspecable.go:168]: The `resource` package is already quite dense. For example, we already have this:

https://github.com/triggermesh/triggermesh/blob/1e3bc1a29fc27455a55a31a8dcf2b2c9bb43aaf7/pkg/reconciler/resource/container.go#L203-L215

For things which aren't meant to be used by individual reconciler implementations but rather by a specific piece of code, I would encourage creating _local_ `ObjectOptions` instead.
In other words, move it to `pkg/reconciler/adapter.go` if you really need it, or use `resource.Requests` and `resource.Limits` instead to avoid having to maintain code that does one thing in different ways.

File: pkg/reconciler/adapter.go [hunk 6]
```diff
@@ -309,3 +317,22 @@ func MaybeAppendValueFromEnvVar(envs []corev1.EnvVar, key string, valueFrom v1al
...
 	return envs
 }
+
+// adapterOverrideOptions applies adapter override parameters depending on
+// deployment type.
+func adapterOverrideOptions(overrides *v1alpha1.AdapterOverrides) resource.ObjectOption {
+	knServiceVisibility := resource.VisibilityClusterLocal
+	if overrides.Public {
+		knServiceVisibility = resource.VisibilityPublic
+	}
```
Comment (antoineco) [inline pkg/reconciler/adapter.go:327]: Echoing a previous comment: we probably want `overrides.Public` to be optional, otherwise `kubectl get` will always show it, even if users never set it.

And I would also move those bits to the `case *servingv1.Service` because that's the only case where this parameter is relevant.

That logic could become something like:

```golang
func adapterOverrideOptions(overrides *v1alpha1.AdapterOverrides) []resource.ObjectOption {
	var opts []resource.ObjectOption // assumes we are now returning individual options

	if _, ok := object.(*servingv1.Service); ok {
		if pub := overrides.Public; pub != nil && pub {
			append(opts, resource.VisibilityPublic)
		}
	}

	if overrides.Resources != nil {
		// do the Resources magic ...
	}
}
```

File: pkg/reconciler/adapter.go [hunk 7]
```diff
@@ -309,3 +317,22 @@ func MaybeAppendValueFromEnvVar(envs []corev1.EnvVar, key string, valueFrom v1al
...
+	return func(object interface{}) {
+		switch o := object.(type) {
+		case *appsv1.Deployment:
```
Comment (antoineco) [inline pkg/reconciler/adapter.go:330]: Dead switch `case`?

**Review by antoineco — COMMENTED** [review]
Looking really good 👍
I have a few suggestions around making the interface optional and cleaning up the code that assembles the override options, otherwise I think this is mostly good to go.

**Comment by antoineco** [comment]
Personally I'm fine with `spec.adapterOverrides`. It does what it says: overrides the default type's adapter settings.
If we prefer `spec.adapter` that's also fine, but I'm not shocked by the current name.

File: pkg/apis/sources/v1alpha1/awscloudwatch_types.go (hunk 2 above)
Comment (tzununbekov) [reply pkg/apis/sources/v1alpha1/awscloudwatch_types.go]: I put this field at the beginning of the structure because there is another common source field `duckv1.SourceSpec` set as a first-line. Also, some of them seemed to be alphabetically ordered. Anyways, I have moved it to the bottom.

**Comment by tzununbekov** [comment]
@antoineco thanks for the early review, most of the comments are taken care of.
There is an issue with `public` switch - it's based on adding and removing a particular label to Kn Service. The "removing" part of this logic does not play well with deep derivative which ignores the missing label on the expected object. I'm also clarifying the resource requests and limits behavior. Will post the update here as soon as I get something.

**Comment by antoineco** [comment]
@tzununbekov ah right, I see what you mean. Labels are serialized to a Go `map`, so the deep derivative comparison cannot take the order into account (unlike `slices`).

There is a simple way to address this: write a custom [comparison function](https://github.com/triggermesh/triggermesh/blob/05df48fbb4432ecd43641cbb9393ce18e40ebf04/pkg/reconciler/semantic/semantic.go#L43-L45) for `map[string]string` that starts by comparing "special" labels (similarly to how "special" labels are propagated from the parent object to its children).

Example:

```golang
	func(a, b map[string]string) bool { // e.g. metadata.labels
		// special handling of labels that can be either added or removed by
		// TriggerMesh reconcilers.
		removableLabels := []string{network.VisibilityLabelKey}
		for _, l := range removableLabels {
			if a[l] != b[l] {
				return false
			}
		}

		return (conversion.Equalities{}).DeepDerivative(a, b)
	},
```

File: pkg/apis/common/v1alpha1/object_interfaces.go [hunk 8]
```diff
@@ -41,6 +41,11 @@ type Reconcilable interface {
 	GetStatusManager() *StatusManager
 }
 
+type AdapterConfigurable interface {
```
Comment (antoineco) [inline pkg/apis/common/v1alpha1/object_interfaces.go:44]: It would be nice to have a comment on that interface, like in the suggestion below, or anything else that conveys what this interface is useful for.

```suggestion
// AdapterConfigurable is implemented by types that can override the default
// configuration of their receive adapter.
type AdapterConfigurable interface {
```

File: pkg/reconciler/adapter.go [hunk 9]
```diff
@@ -309,3 +322,40 @@ func MaybeAppendValueFromEnvVar(envs []corev1.EnvVar, key string, valueFrom v1al
...
+
+	return opts
+}
+
+func toQuantity(resources corev1.ResourceList) (kres.Quantity, kres.Quantity) {
+	var cpu, memory kres.Quantity
+
+	for k, v := range resources {
+		switch k {
+		case "cpu":
```
Comment (antoineco) [inline pkg/reconciler/adapter.go:354]: nit: Kubernetes has a [constant](https://github.com/triggermesh/triggermesh/blob/7547c4b582e34706014a3aaa5f99d4e48e8050da/pkg/reconciler/resource/container.go#L228-L235) defined for that: `corev1.ResourceCPU`

File: pkg/reconciler/adapter.go [hunk 10]
```diff
@@ -309,3 +322,40 @@ func MaybeAppendValueFromEnvVar(envs []corev1.EnvVar, key string, valueFrom v1al
...
+			cpu = v
+		case "memory":
```
Comment (antoineco) [inline pkg/reconciler/adapter.go:356]: nit: Kubernetes has a [constant](https://github.com/triggermesh/triggermesh/blob/7547c4b582e34706014a3aaa5f99d4e48e8050da/pkg/reconciler/resource/container.go#L228-L235) defined for that: `corev1.ResourceMemory`

**Review by antoineco — APPROVED** [review]
Looking good to me 👍 

Approving with nits.

File: config/300-awscloudwatchlogssource.yaml [hunk 11]
```diff
@@ -52,6 +52,41 @@ spec:
...
+              adapterOverrides:
+                description: Kubernetes object parameters to apply on top of default adapter values.
+                type: object
+                properties:
+                  public:
+                    description: Adapter visibility scope.
+                    type: boolean
+                  resources:
+                    description: 'Compute Resources required by this container.
+                      Cannot be updated. More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/'
```
[the remark this answers is not in the document]
Comment (tzununbekov) [reply config/300-awscloudwatchlogssource.yaml:61]: Not relevant. The description is about Pod's resource requirements that cannot be updated directly on the container. But in the context of Deployment or Kn Service, a rolling update does the thing. Not sure how this comment ended up here, I took the CRD part from the Knative resources. Updating.

File: config/300-awscloudwatchlogssource.yaml [hunk 12]
```diff
@@ -52,6 +52,41 @@ spec:
...
+                      requests:
+                        additionalProperties:
+                          anyOf:
+                          - type: integer
+                          - type: string
+                          pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
+                          x-kubernetes-int-or-string: true
+                        description: 'Requests describes the minimum amount
+                          of compute resources required. If Requests is
+                          omitted for a container, it defaults to Limits
```
[the remark this answers is not in the document]
Comment (tzununbekov) [reply config/300-awscloudwatchlogssource.yaml]: Apparently, this is something Kubernetes does. Checked manually this behavior, and it matches its description. 

https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
```
If a container specifies its own memory limit, but does not specify a memory request, Kubernetes automatically
assigns a memory request that matches the limit. Similarly, if a container specifies its own CPU limit, but does not
specify a CPU request, Kubernetes automatically assigns a CPU request that matches the limit.
```

File: pkg/reconciler/resource/container.go [hunk 13]
```diff
@@ -225,13 +225,17 @@ func resourcesFrom(object interface{}) (resources *corev1.ResourceRequirements)
 	return
 }
 
-func setResources(res *corev1.ResourceList, cpu, mem resource.Quantity) {
+func setResources(res *corev1.ResourceList, cpu, mem *resource.Quantity) {
```
Comment (tzununbekov) [inline pkg/reconciler/resource/container.go:228]: @antoineco I had to set these input types to pointers to distinguish when resource quantities are 0 and when they are not set, otherwise setting resource override for say CPU limit would automatically set the Memory limit to 0, while it should remain undefined.
And one last thing - deep derivative does not seem to react to resource overrides deletion on the desired state 🙁 Is there a way to force semantic comparison on override attributes match states "both ways"?

Comment (antoineco) [reply pkg/reconciler/resource/container.go:228]: Sure, changing the `resource` package to accept pointers is also OK if it doesn't support what we need.
I didn't know that a zero quantity would translate to literally `0`.

Comment (antoineco) [reply pkg/reconciler/resource/container.go:228]: Regarding the deep derivative comparison of resources, I'm a bit torn.
Of course we could define a custom comparison function for all the overridable fields (`corev1.ResourceList` and labels for the time being), but I'm wondering how often we expect people to 

1. Set requests/limits
1. Remove those requests/limits

I have a feeling people who do override requests/limits probably will **modify** those values (maybe), but I can't imagine a lot of cases where they will remove them.
If we think it's worth the effort, I'm all game for having an exception for `corev1.ResourceList`, but the case feels a bit edgy to me.

File: pkg/reconciler/resource/knservice_test.go [hunk 14]
```diff
@@ -43,8 +45,8 @@ func TestNewServiceWithDefaultContainer(t *testing.T) {
 		EnvVar("TEST_ENV2", "val2"),
 		Label("test.label/2", "val2"),
 		ServiceAccount("god-mode"),
-		Requests(resource.MustParse("250m"), resource.MustParse("100Mi")),
-		Limits(resource.MustParse("250m"), resource.MustParse("100Mi")),
+		Requests(&cpuRes, &memRes),
+		Limits(&cpuRes, nil),
```
Comment (antoineco) [inline pkg/reconciler/resource/knservice_test.go:49]: In `deployment_test.go` this is set to

```golang
Requests(&cpuRes, &memRes),
Limits(&cpuRes, &memRes),
```

I don't care which one we go for, be I think we should keep both test cases identical, for consistency and to ensure they both yield the same result given the same input.

File: config/300-awscloudwatchlogssource.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (antoineco) [reply config/300-awscloudwatchlogssource.yaml]: Ahhh thank you, learned something!

File: pkg/reconciler/resource/container.go (hunk 13 above)
Comment (tzununbekov) [reply pkg/reconciler/resource/container.go:228]: I'm fine with leaving it as it is right now until someone reports it as an issue.
